Optical Flow Use Improvements - supersedes #9509

priseborough opened the pull request:
Supersedes https://github.com/PX4/Firmware/pull/9509

Rebase only.

priseborough commented:
Attempting to run  'make posix gazebo_iris_opt_flow' with this PR produces linking errors around the mavlink messages:

In file included from /Users/paul/src/Firmware/Tools/sitl_gazebo/src/gazebo_mavlink_interface.cpp:22:
In file included from /Users/paul/src/Firmware/Tools/sitl_gazebo/include/gazebo_mavlink_interface.h:67:
In file included from /Users/paul/src/Firmware/mavlink/include/mavlink/v2.0/common/mavlink.h:32:
In file included from /Users/paul/src/Firmware/mavlink/include/mavlink/v2.0/common/common.h:30:
In file included from /Users/paul/src/Firmware/mavlink/include/mavlink/v2.0/common/../protocol.h:83:
/Users/paul/src/Firmware/mavlink/include/mavlink/v2.0/common/../mavlink_helpers.h:248:25: warning: taking address of packed member 'checksum' of class or structure '__mavlink_message' may result in an unaligned pointer value [-Waddress-of-packed-member]
        crc_accumulate_buffer(&msg->checksum, _MAV_PAYLOAD(msg), msg->len);
                               ^~~~~~~~~~~~~

Flight testing in optical flow showed the upper flow height limiting allowed up to 1.5m of overshoot when the vehicle approached with full climb rate from below, presumably because of the constant acceleration braking assumption. There were also noticeable transients in height as the limiter switched between vertical velocity and position control.

Whilst the behaviour is not unsafe and the limiter performs the essential function of protecting optical flow navigation, the transients and overshoot would unlikely be acceptable in a product. The options are we either:

1)  Accept the deficiencies in limiter behaviour  and rework the height limiter in a follow up PR.

or

2) Rework the height limiter which will require the current issues with running. The SITL gazebo build issues will need to be resolved first.

priseborough commented:
I have tried reinstalling the simulation toolchain following the instructions here: https://dev.px4.io/en/setup/dev_env_mac.html , but some of the packages cannot be found:

==> Reinstalling px4/px4/px4-sim 
==> Installing dependencies for px4/px4/px4-sim: ignition-math3, ignition-tools, ignition-msgs0, zeromq, ignition-transport3, libzzip, ogre1.9, protobuf-c, qt, sdformat5, tinyxml2, bullet, gettext, libffi, pcre, glib, jasper, netpbm, gts, simbody, osrf/simulation/gazebo8
==> Installing px4/px4/px4-sim dependency: ignition-math3
==> Downloading http://gazebosim.org/distributions/ign-math/releases/ignition-math3-3.3.0.sierra.bottle.tar.gz

curl: (22) The requested URL returned error: 500 Internal Server Error
Error: Failed to download resource "ignition-math3"
Download failed: http://gazebosim.org/distributions/ign-math/releases/ignition-math3-3.3.0.sierra.bottle.tar.gz
Warning: Bottle installation failed: building from source.
==> Downloading http://gazebosim.org/distributions/ign-math/releases/ignition-math3-3.3.0.tar.bz2

curl: (22) The requested URL returned error: 500 Internal Server Error
Error: Failed to download resource "ignition-math3"
Download failed: http://gazebosim.org/distributions/ign-math/releases/ignition-math3-3.3.0.tar.bz2

priseborough commented:
Getting SITL working was not achievable in the short  term, however I was able to implement and briefly flight test an improvement to the maximum height limiter https://github.com/PX4/Firmware/pull/9613/commits/3adf8523d1a965b0c5bf9db089f0c29efdc060ff that solves the problem of the height overshoots and transients caused by alt control switching seen previously.

Log here: https://logs.px4.io/plot_app?log=fbba4c54-af36-4f45-bd02-d950f358c8b0

The following plot (signs adjusted so up is positive) shows the effect of the new limiter. The vehicle can no longer be flown up to the altitude limit faster than the altitude control loop can capture it. The height control loops on this vehicle was tuned fairly softly. Tuning the MPC_Z_P gain higher would result in later intervention and a faster climb rate up to the height limit.

![screen shot 2018-06-07 at 2 54 56 pm](https://user-images.githubusercontent.com/3596952/41078936-fb8bfd0a-6a62-11e8-8abe-853adbd2fdc0.png)

@Stifael FYI

Stifael commented:
I think things are a bit too complicated. 
i suggest to just limit the upward velocity close to the limit, similar how it is done for landing and takeoff using the gradual-velocity function (https://github.com/PX4/Firmware/blob/master/src/modules/mc_pos_control/mc_pos_control_main.cpp#L2409).

priseborough commented:
Do you mean always leave it in velocity mode?

priseborough commented:
If we were to leave it in velocity mode, the derivative from distance to limit to vertical velocity lower limit would need to be consistent with the MPC_Z_P gain and we would need to ensure that if above the limit the velocity demand is limited appropriately. This was why i elected to use the altitude control mode so that the dynamic response of the limiter would be consistent with normal operation and no further check would be required.

LorenzMeier commented:
Does your SITL setup issue still persist? Please run `make distclean` and `git submodules update --recursive` and then build again.

priseborough commented:
@Stifael I've implemented the limiter without using the altitude controller as you suggested and pushed the patch to this PR. It flight tested OK, but I am unsure about what protections in terms of acceleration and velocity limits we are bypassing by doing it this way. I added protection for excessive downwards velocity, but that was all.

LorenzMeier commented:
@RomanBapst Could you please follow up here?

priseborough commented:
@LorenzMeier I had already tried that. I will liaase with @RomanBapst  on the issue.

priseborough commented:
@Stifael does the last commit https://github.com/PX4/Firmware/pull/9613/commits/ba3e457f699a7e84c04866510f85cc33ebd0e467 look OK to you?

mhkabir commented on src/modules/mavlink/mavlink_receiver.cpp:
```diff
@@ -2312,8 +2312,10 @@ MavlinkReceiver::handle_message_hil_state_quaternion(mavlink_message_t *msg)
+		_hil_local_pos.vxy_max = INFINITY;
+		_hil_local_pos.vz_max = INFINITY;
+		_hil_local_pos.hagl_min = INFINITY;
+		_hil_local_pos.hagl_max = INFINITY;
```
These should be set to INFINITY.

mhkabir commented on src/modules/simulator/simulator_mavlink.cpp:
```diff
@@ -525,7 +525,9 @@ void Simulator::handle_message(mavlink_message_t *msg, bool publish)
-			hil_lpos.limit_hagl = false;
+			hil_lpos.vz_max = 0.0f;
+			hil_lpos.hagl_min = 0.0f;
+			hil_lpos.hagl_max = 0.0f;
```
These should be set to INFINITY.

RomanBapst commented:
@priseborough I was able to test this in SITL with good results.
It never allowed me to fly higher than SENS_FLOW_MAXHGT.
I also verified that the controller would reduce the height setpoint if SENS_FLOW_MAXHGT was changed during flight.

Here's a plot of the local height and local height setpoint:
![height_limiter](https://user-images.githubusercontent.com/7610489/41158042-df62136a-6b27-11e8-943a-9c2fbafb69bf.png)

Here is the full log:
https://logs.px4.io/plot_app?log=ac93b4e1-c72c-4cd7-ba4b-72b03d02325c

dagar commented:
The ECL side hasn't been merged yet. PX4/Firmware master is currently pointing at a pr-ekfOptFlowFixes.
